File: server-common/src/main/java/su/plo/voice/server/config/VoiceServerLanguages.java
package su.plo.voice.server.config;

import com.google.common.base.Charsets;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.apache.logging.log4j.LogManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import su.plo.config.toml.Toml;
import su.plo.config.toml.TomlWriter;
import su.plo.crowdin.PlasmoCrowdinLib;
import su.plo.voice.api.server.config.ResourceLoader;
import su.plo.voice.api.server.config.ServerLanguages;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.Map;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public final class VoiceServerLanguages implements ServerLanguages {

    private final String defaultLanguageName;

    private final Map<String, VoiceServerLanguage> languages = Maps.newHashMap();

    @Override
    public synchronized void register(@NotNull ResourceLoader resourceLoader,
                                      @NotNull File languagesFolder) {
        try {
            Map<String, VoiceServerLanguage> languages = Maps.newHashMap();

            // load from languages/list
            try (InputStream is = resourceLoader.load("languages/list");
                 BufferedReader br = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))
            ) {
                for (String languageName : br.lines().filter(line -> !line.isEmpty()).collect(Collectors.toList())) {
                    File languageFile = new File(languagesFolder, String.format("%s.toml", languageName));

                    VoiceServerLanguage language = loadLanguage(resourceLoader, languageFile, languageName);
                    languages.put(languageName, language);
                }
            }

            register(languages, languagesFolder);
        } catch (IOException e) {
            throw new IllegalStateException("Failed to load languages", e);
        }
    }

    @Override
    public void register(@NotNull String crowdinProjectId,
                         @Nullable String fileName,
                         @NotNull ResourceLoader resourceLoader,
                         @NotNull File languagesFolder) {
        try {
            try {
                downloadCrowdinTranslations(crowdinProjectId, fileName, languagesFolder);
            } catch (Exception e) {
                LogManager.getLogger().warn(
                        "Failed to download crowdin project {} translations: {}",
                        crowdinProjectId,
                        e.getMessage()
                );
                e.printStackTrace();
            }

            File crowdinFolder = new File(languagesFolder, ".crowdin");
            if (!crowdinFolder.exists()) {
                register(resourceLoader, languagesFolder);
                return;
            }

            Map<String, VoiceServerLanguage> languages = Maps.newHashMap();

            // load from languages/list
            try (InputStream is = resourceLoader.load("languages/list");
                 BufferedReader br = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))
            ) {
                for (String languageName : br.lines().filter(line -> !line.isEmpty()).collect(Collectors.toList())) {
                    String languageFileName = String.format("%s.toml", languageName);

                    File crowdinFile = new File(crowdinFolder, languageFileName);
                    File languageFile = new File(languagesFolder, languageFileName);

                    VoiceServerLanguage language = loadLanguage(resourceLoader, crowdinFile, languageFile, languageName);
                    languages.put(languageName, language);
                }
            }

            // merge defaults from jar and from crowdin, map based on crowdin
            File[] crowdinTranslations = crowdinFolder.listFiles();
            if (crowdinTranslations != null) {
                for (File crowdinFile : crowdinTranslations) {
                    if (crowdinFile.getName().equals("timestamp")) continue;

                    String languageName = crowdinFile.getName().split("\\.")[0];
                    if (languages.containsKey(languageName)) continue;

                    File languageFile = new File(languagesFolder, crowdinFile.getName());

                    VoiceServerLanguage language = loadLanguage(resourceLoader, crowdinFile, languageFile, languageName);
                    languages.put(languageName, language);
                }
            }

            register(languages, languagesFolder);
        } catch (IOException e) {
            throw new IllegalStateException("Failed to load languages", e);
        }
    }

    @Override
    public Map<String, String> getServerLanguage(@Nullable String languageName) {
        return getLanguage(languageName, "server");
    }

    @Override
    public Map<String, String> getClientLanguage(@Nullable String languageName) {
        return getLanguage(languageName, "client");
    }

    private void downloadCrowdinTranslations(@NotNull String crowdinProjectId,
                                             @Nullable String fileName,
                                             @NotNull File languagesFolder) throws Exception {
        long timestamp = 0L;
        File crowdinFolder = new File(languagesFolder, ".crowdin");
        File timestampFile = new File(crowdinFolder, "timestamp");
        if (timestampFile.exists()) {
            String timestampString = new String(Files.readAllBytes(timestampFile.toPath()));
            try {
                timestamp = Long.parseLong(timestampString);
            } catch (NumberFormatException ignored) {
            }
        }

        // check timestamp, if outdated, download from crowdin and use it as defaults
        if (System.currentTimeMillis() - timestamp < 86_400 * 3 * 1000) return;

        Map<String, byte[]> rawTranslations = PlasmoCrowdinLib.INSTANCE
                .downloadRawTranslations(crowdinProjectId, fileName)
                .get();

        // write timestamp file
        crowdinFolder.mkdirs();
        Files.write(
                timestampFile.toPath(),
                String.valueOf(System.currentTimeMillis()).getBytes()
        );

        if (rawTranslations.isEmpty()) return;

        // write translations files
        for (Map.Entry<String, byte[]> entry : rawTranslations.entrySet()) {
            String languageName = entry.getKey();
            byte[] translationBytes = entry.getValue();

            Files.write(
                    new File(crowdinFolder, String.format("%s.toml", languageName)).toPath(),
                    translationBytes
            );
        }
    }

    private void register(@NotNull Map<String, VoiceServerLanguage> languages,
                          @NotNull File languagesFolder) throws IOException {
        if (languages.size() == 0) return;

        final VoiceServerLanguage defaultLanguage = languages.getOrDefault(
                defaultLanguageName,
                languages.get(languages.keySet().iterator().next())
        );

        // load from languagesFolder if not found in list and use default language as defaults
        languagesFolder.mkdirs();

        for (File file : languagesFolder.listFiles()) {
            if (file.isDirectory()) continue;

            String fileName = file.getName();
            if (!fileName.endsWith(".toml")) continue;

            String languageName = fileName.substring(0, fileName.length() - 5);

            if (languages.containsKey(languageName)) continue;

            VoiceServerLanguage language = loadLanguage(file, null);
            if (defaultLanguage != null) language.merge(defaultLanguage);

            languages.put(languageName, language);
        }

        // save all languages
        for (Map.Entry<String, VoiceServerLanguage> entry : languages.entrySet()) {
            File languageFile = new File(languagesFolder, entry.getKey() + ".toml");
            saveLanguage(languageFile, entry.getValue());
        }

        // merge languages
        for (Map.Entry<String, VoiceServerLanguage> entry : languages.entrySet()) {
            VoiceServerLanguage language = this.languages.computeIfAbsent(entry.getKey(), (key) -> entry.getValue());
            language.merge(entry.getValue());
            language.merge(defaultLanguage);
        }

        this.languages.forEach((languageName, language) -> {
            if (languages.containsKey(languageName)) return;
            language.merge(defaultLanguage);
        });
    }

    private Map<String, String> getLanguage(@Nullable String languageName, @NotNull String scope) {
        VoiceServerLanguage language = languages.get(languageName == null ? defaultLanguageName : languageName.toLowerCase());
        if (languageName == null && language == null) return ImmutableMap.of();
        if (language == null) return getLanguage(null, scope);

        return scope.equals("server") ? language.getServerLanguage() : language.getClientLanguage();
    }

    private VoiceServerLanguage loadLanguage(@NotNull ResourceLoader resourceLoader,
                                             @NotNull File crowdinTranslation,
                                             @NotNull File languageFile,
                                             @NotNull String languageName) throws IOException {
        // defaults are based on crowdin translations then merged with defaults from jar (if exist)
        Toml crowdinDefaults = null;
        if (crowdinTranslation.exists()) {
            try (FileInputStream fis = new FileInputStream(crowdinTranslation)) {
                crowdinDefaults = new Toml().read(new InputStreamReader(fis, Charsets.UTF_8));
            }
        }

        Toml jarDefaults = null;
        try (InputStream is = resourceLoader.load("languages/" + languageName + ".toml")) {
            if (is != null) jarDefaults = new Toml().read(new InputStreamReader(is, Charsets.UTF_8));
        } catch (Exception ignored) {
        }

        if (crowdinDefaults == null && jarDefaults == null) {
            throw new IOException("Both crowdin and jar defaults are null for language " + languageName);
        }

        Toml tomlLanguage = new Toml();
        if (languageFile.exists()) {
            try {
                tomlLanguage.read(languageFile);
            } catch (Exception e) {
                throw new IOException("Failed to load language " + languageFile.getName(), e);
            }
        }

        VoiceServerLanguage language = new VoiceServerLanguage(tomlLanguage, null);
        if (crowdinDefaults != null) language.merge(new VoiceServerLanguage(crowdinDefaults, null));
        if (jarDefaults != null) language.merge(new VoiceServerLanguage(jarDefaults, null));

        return language;
    }

    private VoiceServerLanguage loadLanguage(@NotNull ResourceLoader resourceLoader,
                                             @NotNull File languageFile,
                                             @NotNull String languageName) throws IOException {
        try (InputStream is = resourceLoader.load("languages/" + languageName + ".toml")) {
            Toml defaults = new Toml().read(new InputStreamReader(is, Charsets.UTF_8));
            return loadLanguage(languageFile, defaults);
        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException("Failed to load language " + languageName, e);
        }
    }

    private VoiceServerLanguage loadLanguage(@NotNull File languageFile,
                                             @Nullable Toml defaults) throws IOException {
        Toml language = new Toml();
        if (languageFile.exists()) {
            try {
                language.read(languageFile);
            } catch (Exception e) {
                throw new IOException("Failed to load language " + languageFile.getName(), e);
            }
        }

        return new VoiceServerLanguage(language, defaults);
    }

    private void saveLanguage(@NotNull File languageFile,
                              @NotNull VoiceServerLanguage language) throws IOException {
        try {
            new TomlWriter().write(language.getOriginal(), languageFile);
        } catch (Exception e) {
            throw new IOException("Failed to save language", e);
        }
    }

    private static class VoiceServerLanguage {

        @Getter
        private Map<String, Object> original;
        @Getter
        private final Map<String, String> serverLanguage;
        @Getter
        private final Map<String, String> clientLanguage;

        public VoiceServerLanguage(@NotNull Toml language,
                                   @Nullable Toml defaults) {
            this.original = defaults == null
                    ? language.toMap()
                    : mergeMaps(language.toMap(), defaults.toMap());

            this.serverLanguage = mergeMaps(language, "server", defaults);
            this.clientLanguage = mergeMaps(language, "client", defaults);
        }

        public void merge(@NotNull VoiceServerLanguage language) {
            this.original = mergeMaps(original, language.original);

            language.serverLanguage.forEach(serverLanguage::putIfAbsent);
            language.clientLanguage.forEach(clientLanguage::putIfAbsent);
        }

        private Map<String, Object> mergeMaps(@NotNull Map<String, Object> language,
                                              @NotNull Map<String, Object> defaults) {
            Map<String, Object> merged = Maps.newConcurrentMap();
            merged.putAll(language);

            for (Map.Entry<String, Object> entry : defaults.entrySet()) {
                if (entry.getValue() instanceof Map) {
                    merged.put(entry.getKey(), mergeMaps(
                            (Map<String, Object>) language.getOrDefault(entry.getKey(), Maps.newHashMap()),
                            (Map<String, Object>) entry.getValue()
                    ));
                } else if (!language.containsKey(entry.getKey())) {
                    merged.put(entry.getKey(), entry.getValue());
                }
            }

            return merged;
        }

        private Map<String, String> mergeMaps(@NotNull Toml language,
                                              @NotNull String scope,
                                              @Nullable Toml defaults) {
            Map<String, String> defaultsMap = Maps.newConcurrentMap();
            if (defaults != null) {
                defaultsMap.putAll(languageToMapOfStrings(
                        defaults.getTable(scope) == null ? new Toml() : defaults.getTable(scope)
                ));
            }

            Map<String, String> languageMap = languageToMapOfStrings(
                    language.getTable(scope) == null ? new Toml() : language.getTable(scope)
            );

            defaultsMap.putAll(languageMap);
            return defaultsMap;
        }

        private Map<String, String> languageToMapOfStrings(@NotNull Toml language) {
            Map<String, String> languageMap = Maps.newHashMap();

            language.toMap().forEach((key, value) -> {
                if (value instanceof Map) {
                    Map<String, String> tableContents = languageToMapOfStrings(language.getTable(key));

                    tableContents.forEach((contentKey, contentValue) -> {
                        languageMap.put(key + "." + contentKey, contentValue);
                    });
                } else {
                    languageMap.put(key, value.toString());
                }
            });

            return languageMap;
        }
    }
}

File: server-proxy-common/src/main/kotlin/su/plo/voice/server/language/VoiceServerLanguages.kt
package su.plo.voice.server.language

import com.google.common.base.Charsets
import com.google.common.collect.ImmutableMap
import com.google.common.collect.Maps
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.future
import su.plo.config.toml.Toml
import su.plo.config.toml.TomlWriter
import su.plo.crowdin.CrowdinLib
import su.plo.slib.api.language.ServerTranslator
import su.plo.slib.api.logging.McLoggerFactory
import su.plo.voice.api.server.resource.ResourceLoader
import su.plo.voice.api.server.language.ServerLanguages
import java.io.*
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.util.*
import java.util.concurrent.CompletableFuture

class VoiceServerLanguages(
    private val serverTranslator: ServerTranslator,
    override var crowdinEnabled: Boolean = true
) : ServerLanguages {

    private val languages: MutableMap<String, VoiceServerLanguage> = Maps.newHashMap()

    @Synchronized
    override fun register(
        resourceLoader: ResourceLoader,
        languagesFolder: File
    ): CompletableFuture<Void> {
        try {
            val languages: MutableMap<String, VoiceServerLanguage> = Maps.newHashMap()

            val languagesList = readLanguagesList(resourceLoader)
            languagesList.forEach { languageName ->
                val languageFile = File(languagesFolder, String.format("%s.toml", languageName))
                val language = loadLanguage(resourceLoader, languageFile, languageName)
                languages[languageName] = language
            }

            return CompletableFuture.runAsync {
                register(languages, languagesFolder)
            }
        } catch (e: IOException) {
            throw IllegalStateException("Failed to load languages", e)
        }
    }

    override fun register(
        crowdinProjectId: String,
        fileName: String?,
        resourceLoader: ResourceLoader,
        languagesFolder: File
    ): CompletableFuture<Void?> = CoroutineScope(Dispatchers.Default).future {
        registerSync(crowdinProjectId, fileName, resourceLoader, languagesFolder)
        null
    }

    override fun getServerLanguage(languageName: String?) =
        getLanguage(languageName, LanguageScope.SERVER)

    override fun getClientLanguage(languageName: String?) =
        getLanguage(languageName, LanguageScope.CLIENT)

    private fun registerSync(
        crowdinProjectId: String,
        fileName: String?,
        resourceLoader: ResourceLoader,
        languagesFolder: File
    ) {
        try {
            if (crowdinEnabled) {
                try {
                    downloadCrowdinTranslations(crowdinProjectId, fileName, languagesFolder)
                } catch (e: Exception) {
                    LOGGER.warn(
                        "Failed to download crowdin project {} ({}) translations: {}",
                        crowdinProjectId,
                        fileName,
                        e.message
                    )
                }
            }

            val crowdinFolder = File(languagesFolder, ".crowdin")
            if (!crowdinFolder.exists()) {
                register(resourceLoader, languagesFolder).get()
                return
            }

            val languages: MutableMap<String, VoiceServerLanguage> = Maps.newHashMap()

            // load from languages/list
            val languagesList = readLanguagesList(resourceLoader)
            languagesList.forEach { languageName ->
                val languageFileName = "$languageName.toml"

                val crowdinFile = File(crowdinFolder, languageFileName)
                val languageFile = File(languagesFolder, languageFileName)

                val language = loadLanguage(resourceLoader, crowdinFile, languageFile, languageName)
                languages[languageName] = language
            }

            // merge defaults from jar and from crowdin, map based on crowdin
            val crowdinTranslations = crowdinFolder.listFiles()
            if (crowdinTranslations != null) {
                for (crowdinFile in crowdinTranslations) {
                    if (crowdinFile.name == "timestamp") continue

                    val languageName = crowdinFile.name.split(".")[0]
                    if (languages.containsKey(languageName)) continue

                    val languageFile = File(languagesFolder, crowdinFile.name)

                    val language = loadLanguage(resourceLoader, crowdinFile, languageFile, languageName)
                    languages[languageName] = language
                }
            }

            register(languages, languagesFolder)
        } catch (e: IOException) {
            throw IllegalStateException("Failed to load languages", e)
        }
    }

    @Throws(Exception::class)
    private fun downloadCrowdinTranslations(
        crowdinProjectId: String,
        fileName: String?,
        languagesFolder: File
    ) {
        var timestamp = 0L
        val crowdinFolder = File(languagesFolder, ".crowdin")
        val timestampFile = File(crowdinFolder, "timestamp")
        if (timestampFile.exists()) {
            val timestampString = String(Files.readAllBytes(timestampFile.toPath()))
            try {
                timestamp = timestampString.toLong()
            } catch (ignored: NumberFormatException) {
            }
        }

        // check timestamp, if outdated, download from crowdin and use it as defaults
        if (System.currentTimeMillis() - timestamp < 86400 * 3 * 1000) return

        val rawTranslations: Map<String, ByteArray> = CrowdinLib
            .downloadRawTranslations(crowdinProjectId, fileName)
            .get()

        // write timestamp file
        crowdinFolder.mkdirs()
        Files.write(
            timestampFile.toPath(),
            System.currentTimeMillis().toString().toByteArray()
        )

        if (rawTranslations.isEmpty()) return

        // write translations files
        for ((languageName, translationBytes) in rawTranslations) {
            Files.write(
                File(crowdinFolder, "$languageName.toml").toPath(),
                translationBytes
            )
        }
    }

    @Synchronized
    @Throws(IOException::class)
    private fun register(
        languages: MutableMap<String, VoiceServerLanguage>,
        languagesFolder: File
    ) {
        if (languages.isEmpty()) return

        val defaultLanguage = languages.getOrDefault(
            serverTranslator.defaultLanguage,
            languages[languages.keys.first()]!!
        )

        // load from languagesFolder if not found in list and use default language as defaults
        languagesFolder.mkdirs()

        for (file in languagesFolder.listFiles()!!) {
            if (file.isDirectory) continue

            val fileName = file.name
            if (!fileName.endsWith(".toml")) continue

            val languageName = fileName.substring(0, fileName.length - 5)

            if (languages.containsKey(languageName)) continue

            val language = loadLanguage(file, null)
            language.merge(defaultLanguage)

            languages[languageName] = language
        }

        // save all languages
        for ((key, value) in languages) {
            val languageFile = File(languagesFolder, "$key.toml")
            saveLanguage(languageFile, value)
        }

        // merge languages
        for ((languageName, value) in languages) {
            val language = this.languages.computeIfAbsent(languageName) { value }
            language.merge(value)
            language.merge(defaultLanguage)
        }

        this.languages.forEach { (languageName, language) ->
            if (languages.containsKey(languageName)) return@forEach
            language.merge(defaultLanguage)
        }

        this.languages.forEach { (languageName, language) ->
            serverTranslator.register(languageName, language.serverLanguage)
        }
    }

    private fun getLanguage(languageName: String?, scope: LanguageScope): Map<String, String> {
        val language = languages[languageName?.lowercase() ?: serverTranslator.defaultLanguage]
        if (languageName == null && language == null) return ImmutableMap.of()
        if (language == null) return getLanguage(null, scope)

        return if (scope == LanguageScope.SERVER) language.serverLanguage else language.clientLanguage
    }

    @Throws(IOException::class)
    private fun loadLanguage(
        resourceLoader: ResourceLoader,
        crowdinTranslation: File,
        languageFile: File,
        languageName: String
    ): VoiceServerLanguage {
        // defaults are based on crowdin translations then merged with defaults from jar (if exist)
        var crowdinDefaults: Toml? = null
        if (crowdinTranslation.exists()) {
            FileInputStream(crowdinTranslation).use { inputStream ->
                val inputStreamReader = InputStreamReader(inputStream, Charsets.UTF_8)
                crowdinDefaults = Toml().read(inputStreamReader)
            }
        }

        var jarDefaults: Toml? = null
        try {
            resourceLoader.load("languages/$languageName.toml")
                .use { inputStream ->
                    if (inputStream == null) return@use

                    val inputStreamReader = InputStreamReader(inputStream, Charsets.UTF_8)
                    jarDefaults = Toml().read(inputStreamReader)
                }
        } catch (ignored: Exception) {
        }

        if (crowdinDefaults == null && jarDefaults == null) {
            throw IOException("Both crowdin and jar defaults are null for language $languageName")
        }

        val tomlLanguage = Toml()
        if (languageFile.exists()) {
            try {
                tomlLanguage.read(languageFile)
            } catch (e: Exception) {
                throw IOException("Failed to load language " + languageFile.name, e)
            }
        }

        val language = VoiceServerLanguage(tomlLanguage, null)
        if (crowdinDefaults != null) language.merge(VoiceServerLanguage(crowdinDefaults!!, null))
        if (jarDefaults != null) language.merge(VoiceServerLanguage(jarDefaults!!, null))

        return language
    }

    @Throws(IOException::class)
    private fun loadLanguage(
        resourceLoader: ResourceLoader,
        languageFile: File,
        languageName: String
    ): VoiceServerLanguage {
        try {
            resourceLoader.load("languages/$languageName.toml").use { inputStream ->
                if (inputStream == null) throw IOException("Resource 'languages/$languageName.toml' not found")
                val inputStreamReader = InputStreamReader(inputStream, Charsets.UTF_8)

                val defaults: Toml = Toml().read(inputStreamReader)
                return loadLanguage(languageFile, defaults)
            }
        } catch (e: IOException) {
            throw e
        } catch (e: Exception) {
            throw IOException("Failed to load language $languageName", e)
        }
    }

    @Throws(IOException::class)
    private fun loadLanguage(
        languageFile: File,
        defaults: Toml?
    ): VoiceServerLanguage {
        val language = Toml()
        if (languageFile.exists()) {
            try {
                language.read(languageFile)
            } catch (e: Exception) {
                throw IOException("Failed to load language " + languageFile.name, e)
            }
        }

        return VoiceServerLanguage(language, defaults)
    }

    @Throws(IOException::class)
    private fun saveLanguage(
        languageFile: File,
        language: VoiceServerLanguage
    ) {
        try {
            TomlWriter().write(language.original, languageFile)
        } catch (e: Exception) {
            throw IOException("Failed to save language", e)
        }
    }

    private fun readLanguagesList(resourceLoader: ResourceLoader): List<String> {
        resourceLoader.load("languages/list")
            .use { inputStream ->
                val languagesList = inputStream ?: throw IllegalStateException("Resource 'languages/list' not found")

                BufferedReader(InputStreamReader(languagesList, StandardCharsets.UTF_8))
                    .use { br ->
                        return br.readLines().filter { it.isNotEmpty() }
                    }
            }
    }

    private enum class LanguageScope {
        CLIENT,
        SERVER
    }

    private class VoiceServerLanguage(
        language: Toml,
        defaults: Toml?
    ) {

        var original: Map<String, Any>
        val serverLanguage: MutableMap<String, String>
        val clientLanguage: MutableMap<String, String>

        init {
            original = if (defaults == null) language.toMap() else mergeMaps(language.toMap(), defaults.toMap())
            serverLanguage = mergeMaps(language, "server", defaults)
            clientLanguage = mergeMaps(language, "client", defaults)
        }

        fun merge(language: VoiceServerLanguage) {
            original = mergeMaps(original, language.original)

            language.serverLanguage.forEach(serverLanguage::putIfAbsent)
            language.clientLanguage.forEach(clientLanguage::putIfAbsent)
        }

        @Suppress("UNCHECKED_CAST")
        private fun mergeMaps(
            language: Map<String, Any>,
            defaults: Map<String, Any>
        ): Map<String, Any> {
            val merged: MutableMap<String, Any> = Maps.newConcurrentMap()
            merged.putAll(language)

            for ((key, value) in defaults) {
                if (value is Map<*, *>) {
                    merged[key] = mergeMaps(
                        (language[key] ?: Maps.newHashMap<String, Any>()) as Map<String, Any>,
                        value as Map<String, Any>
                    )
                } else if (!language.containsKey(key)) {
                    merged[key] = value
                }
            }

            return merged
        }

        private fun mergeMaps(
            language: Toml,
            scope: String,
            defaults: Toml?
        ): MutableMap<String, String> {
            val defaultsMap: MutableMap<String, String> = Maps.newConcurrentMap()
            if (defaults != null) {
                defaultsMap.putAll(
                    languageToMapOfStrings(
                        if (defaults.getTable(scope) == null) Toml() else defaults.getTable(scope)
                    )
                )
            }

            val languageMap = languageToMapOfStrings(
                if (language.getTable(scope) == null) Toml() else language.getTable(scope)
            )

            defaultsMap.putAll(languageMap)
            return defaultsMap
        }

        private fun languageToMapOfStrings(language: Toml): Map<String, String> {
            val languageMap: MutableMap<String, String> = Maps.newHashMap()

            language.toMap().forEach { (key, value) ->
                if (value is Map<*, *>) {
                    val tableContents = languageToMapOfStrings(language.getTable(key))

                    tableContents.forEach { (contentKey, contentValue) ->
                        languageMap["$key.$contentKey"] = contentValue
                    }
                } else {
                    languageMap[key] = value.toString()
                }
            }
            return languageMap
        }
    }

    companion object {
        private val LOGGER = McLoggerFactory.createLogger("VoiceServerLanguages")
    }
}
